src/Catalyst.KBucket/KBucket.cs

```diff
@@ -0,0 +1,503 @@
+                   .Contacts.Add(contact);
+
+            node.Contacts = null; // mark as inner tree node
+
+            // don't split the "far away" node
+            // we check where the local node would end up and mark the other one as
+            // "dontSplit" (i.e. "far away")
+            var detNode = _DetermineNode(node, LocalContactId, bitIndex);
+            var otherNode = node.Left == detNode ? node.Right : node.Left;
+            otherNode.DontSplit = true;
+        }
+
+        /// <summary>
+        ///   Updates the contact selected by the arbiter.
+        /// </summary>
+        /// <remarks>
+        ///   If the selection is our old contact and the candidate is some new contact
+        ///   then the new contact is abandoned (not added).
+        ///
+        ///   If the selection is our old contact and the candidate is our old contact
+        ///   then we are refreshing the contact and it is marked as most recently
+        ///   contacted(by being moved to the right/end of the bucket array).
+        ///   
+        ///   If the selection is our new contact, the old contact is removed and the new
+        ///   contact is marked as most recently contacted.
+        /// </remarks>
+        private void _Update(Bucket<T> node, int index, T contact)
+        {
+            var incumbent = node.Contacts[index];
+            var selection = Arbiter(incumbent, contact);
+
+            // if the selection is our old contact and the candidate is some new
+            // contact, then there is nothing to do
+            if (selection == incumbent && incumbent != contact) return;
+
+            node.Contacts.RemoveAt(index);
+            node.Contacts.Add(selection);
+        }
+
+        /// <summary>
+        ///   Determines whether the id at the bitIndex is 0 or 1.
+        /// </summary>
+        /// <returns>
+        ///   Left leaf if `id` at `bitIndex` is 0, right leaf otherwise
+        /// </returns>
+        /// <remarks>
+        ///   This is an internal method.  It should not be directly called.
+        ///   It is only public for unit testing.
+        /// </remarks>
+        public Bucket<T> _DetermineNode(Bucket<T> node, byte[] id, int bitIndex)
+        {
+            // id's that are too short are put in low bucket (1 byte = 8 bits)
+            // (bitIndex >> 3) finds how many bytes the bitIndex describes
+            // bitIndex % 8 checks if we have extra bits beyond byte multiples
+            // if number of bytes is <= no. of bytes described by bitIndex and there
+            // are extra bits to consider, this means id has less bits than what
+            // bitIndex describes, id therefore is too short, and will be put in low
+            // bucket
+            var bytesDescribedByBitIndex = bitIndex >> 3;
+            var bitIndexWithinByte = bitIndex % 8;
+            if (id.Length <= bytesDescribedByBitIndex && bitIndexWithinByte != 0) return node.Left;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)